### PR TITLE
Added --timeout to allow custom timeouts for command execution

### PR DESCRIPTION
**Pull request by erinxocon**
This pull requests addresses issues #65 and #401. On slower machines, or machines connected to a slow connection, virtualenv may have problems completing within the default 30 seconds.   These changes allow a user to specify something like `pipenv --two --timeout 60` will allow 60 seconds for the virtualenv to be completely created instead of the 30.  I've tested this on my Raspberry Pi 0 which takes 57 seconds to create a virtualenv.  Setting `--timeout 80` let it complete and I'm happily using pipenv to deploy code to the pi zero!  I updated `delegator.py` in the vendor release, and I put a pull request on on the actual delegator.py repo containing the version I'm using here!

### CODE REVIEW

**Comment by kennethreitz** [comment]
Let's just try bumping this to 120 by default, instead of making it configurable.

**Comment by kennethreitz** [comment]
Let's add `PIPENV_TIMEOUT` to `environments.py` and use that as a constant everywhere.

**Comment by kennethreitz** [comment]
related: https://github.com/kennethreitz/delegator.py/pull/25

**Comment by erinxocon** [comment]
Will make changes accordingly!

File: README.rst [hunk 1]
```diff
@@ -22,7 +22,7 @@ Pipenv: Sacred Marriage of Pipfile, Pip, & Virtualenv
 ---------------
 
 
-**Pipenv** harnesses `Pipfile <https://github.com/pypa/pipfile>`_, `Pip <https://github.com/pypa/pip>`_, and `Virtualenv <https://github.com/pypa/virtualenv>`_ together in unison to create a single, high-quality tool that is optimized for workflow efficiency and best practices. 
+**Pipenv** harnesses `Pipfile <https://github.com/pypa/pipfile>`_, `Pip <https://github.com/pypa/pip>`_, and `Virtualenv <https://github.com/pypa/virtualenv>`_ together in unison to create a single, high-quality tool that is optimized for workflow efficiency and best practices.
```
Comment (nateprewitt) [inline README.rst]: Could we move doc cleanup into another PR so we have things properly partitioned? That way if this is ever reverted for some reason we're not also reintroducing issues to the documentation.

File: README.rst [hunk 2]
```diff
@@ -31,7 +31,7 @@ Pipenv automatically creates and manages the virtualenvs of your projects, as we
 The propblems that Pipenv seeks to solve are multi-faceted:
 
 - When using Pipenv, you no longer need to use ``pip`` and ``virtualenv`` separately. They work together.
-- Managing a ``requirements.txt`` file `can be problematic <https://www.kennethreitz.org/essays/a-better-pip-workflow>`_, so Pipenv uses the upcoming ``Pipfile`` and ``Pipfile.lock`` instead, which is superior for basic use cases. 
+- Managing a ``requirements.txt`` file `can be problematic <https://www.kennethreitz.org/essays/a-better-pip-workflow>`_, so Pipenv uses the upcoming ``Pipfile`` and ``Pipfile.lock`` instead, which is superior for basic use cases.
```
Comment (nateprewitt) [inline README.rst]: The should be in the docs cleanup PR too.

File: README.rst [hunk 3]
```diff
@@ -52,7 +52,7 @@ User Testimonials
 ----------
 
 - Enables truly *deterministic builds*, while easily specifying *what you want*.
-- Generates and checks file hashes for locked dependencies (via ``--hashes``). 
+- Generates and checks file hashes for locked dependencies (via ``--hashes``).
```
Comment (nateprewitt) [inline README.rst]: Docs cleanup PR.

File: docs/index.rst [hunk 4]
```diff
@@ -77,6 +77,7 @@ Other Commands
 - ``run`` will run a given command from the virtualenv, with any arguments forwarded (e.g. ``$ pipenv run python``).
 - ``check`` asserts that PEP 508 requirements are being met by the current environment.
 
+
```
Comment (nateprewitt) [inline docs/index.rst]: Docs cleanup PR.

File: pipenv/cli.py [hunk 5]
```diff
@@ -21,7 +21,7 @@
 
 from .project import Project
 from .utils import (convert_deps_from_pip, convert_deps_to_pip, is_required_version,
-    proper_case, pep423_name, split_vcs)
+                    proper_case, pep423_name, split_vcs)
```
Comment (nateprewitt) [inline pipenv/cli.py]: This is actually intentional styling that we've used across many of the Kenneth Reitz projects. I think we should probably leave this how it was.

File: pipenv/cli.py [hunk 6]
```diff
@@ -41,7 +41,8 @@
 #         |_|
 
 # Packages that should be ignored later.
-BAD_PACKAGES = ('setuptools', 'pip', 'wheel', 'six', 'packaging', 'pyparsing', 'appdirs')
+BAD_PACKAGES = ('setuptools', 'pip', 'wheel', 'six',
```
Comment (nateprewitt) [inline pipenv/cli.py]: I don't think we need to split lines on this. We don't have a hard cap at 80.

File: pipenv/cli.py [hunk 7]
```diff
@@ -104,7 +105,8 @@ def ensure_pipfile(validate=True):
 
             # Parse requirements.txt file with Pip's parser.
             # Pip requires a `PipSession` which is a subclass of requests.Session.
-            # Since we're not making any network calls, it's initialized to nothing.
+            # Since we're not making any network calls, it's initialized to
```
Comment (nateprewitt) [inline pipenv/cli.py]: Same with this.

File: pipenv/cli.py [hunk 8]
```diff
@@ -229,7 +231,8 @@ def do_install_dependencies(dev=False, only=False, bare=False, requirements=Fals
     if requirements:
         bare = True
 
-    # Load the lockfile if it exists, or if only is being used (e.g. lock is being used).
+    # Load the lockfile if it exists, or if only is being used (e.g. lock is
```
Comment (nateprewitt) [inline pipenv/cli.py]: Not really a need to new line here either.

File: pipenv/cli.py [hunk 9]
```diff
@@ -272,7 +275,8 @@ def do_install_dependencies(dev=False, only=False, bare=False, requirements=Fals
 
         if c.return_code != 0:
             click.echo(crayons.red('An error occured while installing!'))
-            # We echo both c.out and c.err because pip returns error details on out.
+            # We echo both c.out and c.err because pip returns error details on
```
Comment (nateprewitt) [inline pipenv/cli.py]: Single line.

File: pipenv/cli.py [hunk 10]
```diff
@@ -336,7 +340,8 @@ def parse_install_output(output):
     for section in output_sections:
         lines = section.split('\n')
 
-        # Strip dependency parens from name line. e.g. package (from other_package)
+        # Strip dependency parens from name line. e.g. package (from
```
Comment (nateprewitt) [inline pipenv/cli.py]: Single line.

File: pipenv/cli.py [hunk 11]
```diff
@@ -409,7 +414,7 @@ def parse_download_fname(fname, name):
         fname, _ = os.path.splitext(fname)
 
     # Substring out package name (plus dash) from file name to get version.
-    version = fname[len(name)+1:]
+    version = fname[len(name) + 1:]
```
Comment (nateprewitt) [inline pipenv/cli.py]: Is the linter complaining about this?

File: pipenv/cli.py [hunk 12]
```diff
@@ -580,7 +586,8 @@ def do_init(dev=False, requirements=False, allow_global=False, ignore_hashes=Fal
             cleanup_virtualenv(bare=False)
             sys.exit(1)
 
-    # Write out the lockfile if it doesn't exist, but not if the Pipfile is being ignored
+    # Write out the lockfile if it doesn't exist, but not if the Pipfile is
```
Comment (nateprewitt) [inline pipenv/cli.py]: Probably don't need this on a newline.

File: pipenv/cli.py [hunk 13]
```diff
@@ -634,17 +640,20 @@ def pip_install(package_name=None, r=None, allow_global=False, ignore_hashes=Fal
...
             install_reqs += ' --require-hashes'
 
-        c = delegator.run('"{0}" install {1} -i {2}'.format(which_pip(allow_global=allow_global), install_reqs, source['url']))
+        c = delegator.run('"{0}" install {1} -i {2}'.format(
+            which_pip(allow_global=allow_global), install_reqs, source['url']))
 
         if c.return_code == 0:
             break
-    # return the result of the first one that runs ok or the last one that didn't work
+    # return the result of the first one that runs ok or the last one that
```
Comment (nateprewitt) [inline pipenv/cli.py]: Single line.

File: pipenv/cli.py [hunk 14]
```diff
@@ -704,7 +713,7 @@ def format_pip_error(error):
     error = error.replace('THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE', str(crayons.red('THESE PACKAGES DO NOT MATCH THE HASHES FROM Pipfile.lock!', bold=True)))
     error = error.replace('someone may have tampered with them', str(crayons.red('someone may have tampered with them')))
 
-    error = error.replace('option to pip install', 'option to \'pipenv install\'')
+    error = error.replace('option to pip install','option to \'pipenv install\'')
```
Comment (nateprewitt) [inline pipenv/cli.py]: Why was the space removed here?

File: pipenv/cli.py [hunk 15]
```diff
@@ -809,15 +821,15 @@ def install(package_name=False, more_packages=False, dev=False, three=False, pyt
...
+    ensure_project(three=three, python=python, timeout=timeout)
 
     # Capture -e argument and assign it to following package_name.
     more_packages = list(more_packages)
     if package_name == '-e':
         package_name = ' '.join([package_name, more_packages.pop(0)])
 
     # Allow more than one package to be provided.
-    package_names = [package_name,] + more_packages
+    package_names = [package_name, ] + more_packages
```
Comment (nateprewitt) [inline pipenv/cli.py]: I don't think we want a space here.

File: pipenv/cli.py [hunk 16]
```diff
@@ -916,7 +928,8 @@ def uninstall(package_name=False, more_packages=False, three=None, python=False,
         if pipfile_remove:
             norm_name = pep423_name(package_name)
             if norm_name in project._pipfile.get('dev-packages', {}) or norm_name in project._pipfile.get('packages', {}):
-                click.echo('Removing {0} from Pipfile...'.format(crayons.green(package_name)))
+                click.echo('Removing {0} from Pipfile...'.format(
```
Comment (nateprewitt) [inline pipenv/cli.py]: I think this can stay as a single line.

File: pipenv/environments.py [hunk 17]
```diff
@@ -26,7 +26,7 @@
 PIPENV_DEFAULT_THREE = os.environ.get('PIPENV_DEFAULT_THREE')
 
 # Tells Pipenv to use the virtualenv-provided pip instead.
-PIPENV_USE_SYSTEM = os.environ.get('VIRTUAL_ENV') if not 'PIPENV_IGNORE_VIRTUALENVS' in os.environ else False
+PIPENV_USE_SYSTEM = os.environ.get('VIRTUAL_ENV') if 'PIPENV_IGNORE_VIRTUALENVS' not in os.environ else False
```
Comment (nateprewitt) [inline pipenv/environments.py]: This is a good spot! I think it should be submitted separate from this PR though for the same reasons listed above.

**Comment by erinxocon** [comment]
@nateprewitt I'm talking with @kennethreitz now, will make the changes you guys requested and resubmit the PR!

**Comment by nateprewitt** [comment]
@erinxocon great, thanks!

**Comment by erinxocon** [comment]
@nateprewitt VS Code was autoformatting, which I did disable, but not before it could have it's hand at trying to "fix" some things.  That's why there are random new lines and spaces.  I'm going to reset HEAD back a bit and cherry pick things out in regards to functionality, but not style.  Will get everything worked out over today!

**Comment by erinxocon** [comment]
@kennethreitz @nateprewitt squashed a bunch of the commits to all but three for the change.  environments.py now will attempt to look up a environment variable named PIPENV_TIMEOUT, and if it doesn't find it, defaults to 120 seconds.  I tested this on my raspberry pi zero and was successfully able to create a virutalenv via pipenv!  I was going to make a small note in the advanced documentation and submit that in another PR.

File: pipenv/environments.py (hunk 17 above)
Comment (erinxocon) [inline pipenv/environments.py]: this change is still apart of this commit batch, as I had to make changes to environments.py anyways.  Is that ok?

**Comment by kennethreitz** [comment]
sorry the test suite wasn't running

File: pipenv/environments.py [hunk 18]
```diff
@@ -35,3 +35,8 @@
 # Disable spinner on windows.
 if os.name == 'nt':
     PIPENV_NOSPIN = True
+
+_pipenv_timout_raw = os.environ.get('PIPENV_TIMEOUT')
```
Comment (nateprewitt) [inline pipenv/environments.py]: Perhaps we could just set 120 as the fallback here and save the extra lines?

`int(os.environ.get('PIPENV_TIMEOUT', 120)`

File: pipenv/environments.py (hunk 17 above)
Comment (nateprewitt) [inline pipenv/environments.py]: Yeah, we can bundle it in this time. I don't foresee us having an issue with this but a practice to note for the future.

File: pipenv/cli.py [hunk 19]
```diff
@@ -392,7 +392,7 @@ def do_create_virtualenv(three=None, python=None):
 
     # Actually create the virtualenv.
     with spinner():
-        c = delegator.run(cmd, block=False)
+        c = delegator.run(cmd, block=False, binary=False, timeout=PIPENV_TIMEOUT)
```
Comment (nateprewitt) [inline pipenv/cli.py]: Does binary need to be specified here, isn't False the default?

File: pipenv/environments.py (hunk 18 above)
Comment (erinxocon) [inline pipenv/environments.py]: indeed, that is a more concise way to do it, I'll make the change and push it.

File: pipenv/cli.py (hunk 19 above)
Comment (erinxocon) [inline pipenv/cli.py]: No it doesn't I thought it was causing some tests to fail so I added it in, will remove!

**Comment by erinxocon** [comment]
@nateprewitt changes made!  Awaiting your review once again :)

**Comment by nateprewitt** [comment]
Thanks @erinxocon! I just left a note in #406, but if we can get that addressed really quickly, we're ready to merge I think.

**Comment by erinxocon** [comment]
@nateprewitt tests integrated and seem to be passing!

File: tests/test_pipenv.py [hunk 20]
```diff
@@ -66,6 +66,37 @@ def test_cli_usage(self):
...
+        delegator.run('rm -fr test_timeout_long')
+        del os.environ['PIPENV_TIMEOUT']
+
+    def test_timeout_short(self):
+        delegator.run('mkdir test_timeout_short')
+        os.chdir('test_timeout_short')
+
+        os.environ['PIPENV_VENV_IN_PROJECT'] = '1'
+        os.environ['PIPENV_TIMEOUT'] = '1'
+        print(os.environ['PIPENV_TIMEOUT'])
```
Comment (nateprewitt) [inline tests/test_pipenv.py]: Can we get rid of this print statement?

Comment (erinxocon) [inline tests/test_pipenv.py]: @nateprewitt done! Sorry about that

**Comment by nateprewitt** [comment]
Alright, here we go 🚀

Thanks @erinxocon! ✨

**Comment by erinxocon** [comment]
✨🍰✨

**Comment by erinxocon** [comment]
Thanks @nateprewitt for the guidance.  I've been programming for a couple years but haven't contributed to open source software much and decided I wanted to dig in more!  Still getting the hang of practices and what not!

**Comment by nateprewitt** [comment]
@erinxocon, no problem. Thanks for donating time to make the project better!

**Comment by erinxocon** [comment]
@nateprewitt of course!  I'm planning on deploying a project I recently finished for work to a bunch of raspberry pi zeros and I wanted to use pipenv (because it's awesome) for that but it kept failing!  Glad I could contribute and make my life easier now that I can deploy all of my code to the pi and run it easily!  Yay for deterministic builds!

**Comment by jkleve** [comment]
There is no --timeout options in the program usages.

**Comment by habnabit** [comment]
@jkleve apparently it's `PIPENV_TIMEOUT=` in the environment.
